src/__node_tests__/index.js

```diff
@@ -77,6 +85,89 @@ test('works without a browser context on a dom node (JSDOM Fragment)', () => {
   `)
 })
 
+test('works with a custom configured element query for shadow dom elements', async () => {
+  const window = new JSDOM(`
+    <html>
+      <body>
+        <example-input></example-input>
+      </body>
+    </html>
+  `).window
+  const document = window.document
+  const container = document.body
+
+  // create custom element as system under test
+  window.customElements.define(
+    'example-input',
+    class extends window.HTMLElement {
+      constructor() {
+        super()
+        const shadow = this.attachShadow({mode: 'open'})
+
+        const div = document.createElement('div')
+        const label = document.createElement('label')
+        label.setAttribute('for', 'invisible-from-outer-dom')
+        label.innerHTML =
+          'Visible in browser, invisible for traditional queries'
+        const input = document.createElement('input')
+        input.setAttribute('id', 'invisible-from-outer-dom')
+        div.appendChild(label)
+        div.appendChild(input)
+        shadow.appendChild(div)
+      }
+    },
+  )
+
+  // Given the default configuration is used
+
+  // When querying for the label
+  // Then it is not in the document
+  expect(
+    dtl.queryByLabelText(
+      container,
+      /Visible in browser, invisible for traditional queries/i,
+    ),
+  ).not.toBeInTheDocument()
+
+  // Given I have a naive query that allows searching shadow dom
+  const queryMeAndChildrenAndShadow = (element, query) => [
+    ...element.querySelectorAll(query),
+    ...[...element.children].reduce(
+      (result, child) => [
+        ...result,
+        ...queryMeAndChildrenAndShadow(child, query),
+      ],
+      [],
+    ),
+    ...(element.shadowRoot?.querySelectorAll(query) ?? []),
+  ]
```

Comment:
Wouldn't this duplicate elements in e.g. querying for `'button'` in `<div><button /></div>`?

Reply:
It is supposed to be a naive implementation :D 

I added ":scope" to only query children in https://github.com/testing-library/dom-testing-library/pull/1054/commits/28168530dad4e274268a6df78503c9483bc7e413